### ISSUE TITLE
VSMac carret on save refresh whole page to the top.

### ISSUE THREAD

**Issue by bcaceiro** [Xavalon/XamlStyler#311]
**Describe the bug**
On Visual Studio for Mac, having the Format XAML on save option, the carret moves to the top of the current page. this is kind of annoying if working / making changes at the end of a 200 line file, and being prompted to the top of the same.

**To Reproduce**
1. When formatting '...' the carret moves to the sart of the page

**Expected behavior**
The format should be applied and stay at the same editing position

**Version Info (please complete the following information):**
 - OS: MacOS 11.0.1
 - Visual Studio: VS for Mac 8.8.1 
 - XAML Styler: Extension 2.0.1

**Additional context**

**Comment by Saratsin** [comment]
The current logic works like this: 
1. Getting all the text from document
2. Formattimg text
3. Replacing document text with the formatted one.

I can save the cursor position at the point 1. But the problem is that after the formatting the number of symbols before the carret can change dramatically (all the spaces and \n \t). @grochocki can you explain a little how have you fixed this problem in Windows extension?

**Comment by grochocki** [comment]
I could not find anything that handles the cursor position explicitly in the Windows extension. In comparing [XamlFormattingService.cs](https://github.com/Xavalon/XamlStyler/blob/master/src/XamlStyler.Extension.Mac/Services/XamlFormatting/XamlFormattingService.cs) (Mac) and [StylerPackage.Formatting.cs](https://github.com/Xavalon/XamlStyler/blob/master/src/XamlStyler.Extension.Windows/StylerPackage.Formatting.cs) (Windows), my best guess is that it has to do with _how_ we are replacing the text.

We are using [EditPoint](https://docs.microsoft.com/en-us/dotnet/api/envdte.editpoint?view=visualstudiosdk-2019) to edit the text buffer.

> EditPoint objects are similar to TextSelection objects, except that they operate on data in the text buffer rather than on text displayed in a code editor. The difference is that text in the buffer is not affected by global editor states, such as word wrapping and virtual spaces.

And then we use the [vsEPReplaceTextKeepMarkers](https://docs.microsoft.com/en-us/dotnet/api/envdte.vsepreplacetextoptions?view=visualstudiosdk-2019) flag, which:

> Retains existing markers (for breakpoints, named shortcuts, and so forth) in their current locations.

Neither of these mention cursor position explicitly, but if I try to set a different flag, the cursor position jumps to the end of the document on format.

**Comment by IvoKrugers** [comment]
> The current logic works like this:
> 
> 1. Getting all the text from document
> 2. Formattimg text
> 3. Replacing document text with the formatted one.

I understand that the number of lines and their length change drastically at the first save, but after that the changes are subtle.

If you were to save the first visible line and the line where the carrot is at in step 1 and then just position after step 3. That would already improve the experience a lot!

**Comment by ajtruckle** [comment]
If we give elements names, and you not the name and property with focus, you could find it again as you now have context.

**Comment by grochocki** [comment]
Closing this as won't fix since [Visual Studio for Mac is scheduled for retirement by August 31, 2024](https://learn.microsoft.com/en-us/visualstudio/mac/what-happened-to-vs-for-mac?view=vsmac-2022).
